### ISSUE TITLE
Unexpected Generic[] behavior with LiteralString-bounded TypeVar

### ISSUE THREAD

**Issue by brentyi** [microsoft/pyright#4857]
Hi!

I'm seeing some interesting behavior on pyright `1.1.300` when bounding a `TypeVar` with `LiteralString`. Here's a quick example that type checks without errors:
```python
import dataclasses
from typing import Generic, LiteralString, TypeVar

T = TypeVar("T")


@dataclasses.dataclass
class Container(Generic[T]):
    """Dummy generic class."""
    inner: T


TLiteralString = TypeVar("TLiteralString", bound=LiteralString)


def wrap_with_default(
    x: TLiteralString | None,
    default: TLiteralString,
) -> Container[TLiteralString]:
    """Wrap an input with a dummy container object."""
    if x is None:
        x = default  # `x` should now be narrowed to TLiteralString.
    return Container(x)
```

Unexpectedly, this breaks when I simply swap the final return statement for an assignment followed by a return:

```python
import dataclasses
from typing import Generic, LiteralString, TypeVar

T = TypeVar("T")


@dataclasses.dataclass
class Container(Generic[T]):
    """Dummy generic class."""
    inner: T


TLiteralString = TypeVar("TLiteralString", bound=LiteralString)


def wrap_with_default(
    x: TLiteralString | None,
    default: TLiteralString,
) -> Container[TLiteralString]:
    """Wrap an input with a dummy container object."""
    if x is None:
        x = default
    out = Container(x)  # New
    return out  # New
```

Where the final line now gives me the error: `Expression of type "Container[TLiteralString@wrap_with_default | str]" cannot be assigned to return type "Container[TLiteralString@wrap_with_default]"`. In contrast, mypy runs without errors for both examples.

As additional notes, this issue disappears when I:
- Remove the `LiteralString` bound.
- Use `Tuple[TLiteralString]` instead of my custom `Container[TLiteralString]`.

**Comment by erictraut** [comment]
This is working as expected. In your first example, you're using bidirectional type inference to infer the type of the return expression `Container(x)`. In the second example, you're assigning the expression `Container(x)` to a variable with no type annotation, so regular type inference rules are used (without the benefit of bidirectional type inference). If you add an explicit type declaration to the `out` variable, you will see the same behavior as you do in the first example.

```python
out: Container[TLiteralString] = Container(x)
```

**Comment by brentyi** [comment]
Interesting, thanks!

The code I have in practice is closer to the following, where I have a factory method that instantiates `Container`:

```python
import dataclasses
from typing import Generic, TypeVar, Union

from typing_extensions import LiteralString

T = TypeVar("T")


@dataclasses.dataclass
class Container(Generic[T]):
    """Dummy generic class."""

    inner: T


def make_container(x: T) -> Container[T]:  # New
    return Container(x)


TLiteralString = TypeVar("TLiteralString", bound=LiteralString)


def wrap_with_default(
    x: Union[TLiteralString, None],
    default: TLiteralString,
) -> Container[TLiteralString]:
    """Wrap an input with a dummy container object."""
    if x is None:
        x = default
    out: Container[TLiteralString] = make_container(x)  # Error here.
    return out
```

Here I get the error:
`Expression of type "Container[TLiteralString@wrap_with_default | str]" cannot be assigned to declared type "Container[TLiteralString@wrap_with_default]"`.

Is this also as designed? The appearance of the `str` type in the union was unexpected to me, but perhaps that's inevitable given the control flow?

This also works fine if I remove the `LiteralString` bound or replace it with a different bound (like `bound=int` or `bound=str`).

**Comment by brentyi** [comment]
Sounds great, thanks! Appreciate the effort you put into this.

**Comment by erictraut** [comment]
This is included in pyright 1.1.301, which I just published. It will also be included in a future release of pylance.
